draft-ietf-tls-esni.md

```diff
@@ -239,12 +239,12 @@ The `ECHConfigContents` structure contains the following fields:
 key_config
 : A `HpkeKeyConfig` structure carrying the configuration information associated
 with the HPKE public key. Note that this structure contains the `config_id`
-field, which applies to the entire ECHConfigContents. Sites MUST NOT publish
+field, which applies to the entire ECHConfigContents. Sites SHOULD NOT publish
 two different `ECHConfigContents` values with the same `HpkeKeyConfig` value.
 The RECOMMENDED technique for choosing `config_id` is to choose a random byte.
-This process is repeated if this config_id matches that of any valid ECHConfig,
-which could include any ECHConfig that has been recently removed from active
-use.
```

Comment:
This sentence is important to the recommendation, otherwise it doesn't actually avoid collisions.

Reply:
I'm not clear how CDN-A knows that CDN-B (also used by the backend) will in future not collide. So while CDN-A can repeat the process as often as it wants, and the ECHConfig value from CDN-A won't collide with one another, the backend can still get a collision. Or where am I wrong?

Reply:
They don't know, and don't have to know. You only hit trial decryption if, within a set of active configs *on the client-facing server*, there is a config ID collision. The backend doesn't deal with configs in the first place.

Reply:
Sure. trial decryption is fine. The issue is that the sentence I suggest removing seems to me to claim that one can avoid config_id collisions in the DNS if you iterate the random picking. But that's not the case.

Reply:
If one changed that sentence via s/any valid ECHConfig/any known ECHConfig/ it'd be correct I guess.

Reply:
The DNS isn't the scope we care about for collisions in the first place. It sounds like there's some misunderstanding here as to what the config_id is even for. I agree the current text is not right. However, this PR does not fix it. "Any known ECHConfig" is also weird after you've rotated 256 times.

Let's take a step back here: we're encrypting a message for the client-facing server, which then dispatches to a backend server. An ECHConfig is a property of the client-facing server. It's the one generating the keys, etc. In fact, the backend server may not even have the key.

In the DNS, we *typically* only need one ECHConfig per route, because the client is ultimately going to just pick one. The multiplicity in the ECHConfigList is purely so a client-facing server can simultaneously support draft-13 and the final draft, or X25519 and fancy-future-post-quantum-KEM.

However, we never design systems with unrotatable keys. That means, at some point, the client-facing server is going to generate a new a key and update DNS. DNS is cached, TTLs are lies, etc., so old DNS entries may still be used for a while after rotation.

That means the client-facing server needs to support a _window_ of ECHConfigs, perhaps the last week or so's worth, to avoid service interrupts. This window is *larger* than what's in DNS because, again, DNS only really needs the current ones.

When the client-facing server gets a ClientHello, it needs to know which of those to use. It can either trial-decrypt, or it can filter on config_id. Trial decryption means, for every config you keep around, you have to do that many HPKE operations. That is poor performance and a DoS risk. Thus, we have a config_id, enough to disambiguate however many entries the client-facing server needs to keep alive. You *can* reuse config_ids, just, if you want to avoid trial decryption, you must have retired the old one first.

That means, when a client-facing server generates new keys, it should do so in a way that avoids collisions with any existing *active* ones. Rejection sampling is one strategy. Just incrementing and cycling IDs is another. (TBH, cycling seems simpler.)

All of this is relative to one client-facing server. If you use both CDN-A and CDN-B, A and B have separate sets of active keys and separate config ID spaces. There is no need for them to coordinate because we don't care about config_id collisions. CDN-A is never going to try one of CDN-B's keys.

I'll take a stab at rewriting this.

Reply:
I agree with all the above except the misunderstanding bit:-) The "no need for them to co-ordinate" at the end is important IMO and one reason the original "MUST not publish collisions" was wrong. s/MUST/SHOULD/ and s/valid/known active/ would be enough of a fix for me fwiw if you prefer minimalism:-)